lib/src/annotate.rs

```diff
@@ -0,0 +1,397 @@
+}
+
+/// Starting at the starting commit, compute changes at that commit, updating
+/// the mappings. So long as there are mappings left in local_line_map, we
+/// continue. Once local_line_map is empty, we've found sources for each line
+/// and exit.
+fn process_commits(
+    results: &mut PartialResults,
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+    file_name: &RepoPath,
+    num_lines: usize,
+) -> Result<(), AnnotateError> {
+    let predicate = RevsetFilterPredicate::File(FilesetExpression::file_path(file_name.to_owned()));
+    let revset = RevsetExpression::commit(starting_commit_id.clone())
+        .union(
+            &RevsetExpression::commit(starting_commit_id.clone())
+                .ancestors()
+                .filtered(predicate),
+        )
+        .evaluate_programmatic(repo)
+        .map_err(|e| match e {
+            RevsetEvaluationError::StoreError(backend_error) => AnnotateError::from(backend_error),
+            RevsetEvaluationError::Other(_) => {
+                AnnotateError::FileNotFound(file_name.to_internal_dir_string())
+            }
+        })?;
+
+    for (cid, edge_list) in revset.iter_graph() {
+        process_commit(results, repo, file_name, &cid, &edge_list)?;
+        if results.original_line_map.len() >= num_lines {
+            break;
+        }
+    }
+    Ok(())
+}
+
+/// For a given commit, for each parent, we compare the version in the parent
+/// tree with the current version, updating the mappings for any lines in
+/// common. If the parent doesn't have the file, we skip it.
+/// After iterating through all the parents, any leftover lines unmapped means
+/// that those lines are original in the current commit. In that case,
+/// original_line_map is updated for the leftover lines.
+fn process_commit(
+    results: &mut PartialResults,
+    repo: &ReadonlyRepo,
+    file_name: &RepoPath,
+    current_commit_id: &CommitId,
+    edges: &Vec<GraphEdge<CommitId>>,
+) -> Result<(), AnnotateError> {
+    let current_tree = repo.store().get_commit(current_commit_id)?.tree()?;
+    let current_file_id = get_file_id(current_commit_id, &current_tree, file_name)?.unwrap();
+
+    for parent_edge in edges {
+        if parent_edge.edge_type != GraphEdgeType::Missing {
+            let parent_commit = repo.store().get_commit(&parent_edge.target)?;
+            let parent_tree = parent_commit.tree()?;
+            let parent_file_id = get_file_id(parent_commit.id(), &parent_tree, file_name)?;
+
+            if let Some(parent_file_id) = parent_file_id {
+                process_file_ids(
+                    results,
+                    repo.store(),
+                    file_name,
+                    &current_file_id,
+                    current_commit_id,
+                    &parent_file_id,
+                    parent_commit.id(),
+                )?;
+            }
+        }
+    }
+    results.drain_remaining_for_commit_id(current_commit_id);
```

Comment:
Perhaps, this, `process_file_ids()`, `forward_to_new_commit()`, and `drain_remaining_for_commit_id()` can be reorganized in a way that the current line_map entry is popped first.

```rust
let Some(current_map) = self.local_line_map.remove(current_commit_id);
for parent_edge in edges {
    ...
}
results.drain(current_map);
```

It will probably solve the issue of mutable borrows, and clarifies that the `current_map` is consumed.

Reply:
Just in case this comment was buried, the intent is to remove `local_line_map.get_mut(old_commit_id)`/`local_line_map.get_mut(new_commit_id)` hash map lookup inside the lines loop.

Reply:
I'm happy to try out this change but I don't really get the reasoning behind it. Is it an issue to have the double mutation?

Reply:
Yes, that's what I meant, thanks.

It might be better to add `struct { line_map: HashMap<usize, usize>, file_content: Vec<u8> }`. What `process_files_in_commits()` will do is basically forwarding `current_commit.line_map` to `parent_commit.line_map` by comparing `current_commit.file_content` and `parent_commit.file_content`.

Reply:
Ok, I pushed another iteration. Along your lines, PartialResults has been split out since it was slowly disappearing. 
In this iteration, I removed many instances where was pass &mut references and instead the return the results of each commit's computation which are then compiled into hashmaps close to the top level. 
With these changes, I removed Partial Results completely and just pass the necessary objects. Type aliases in the code help differentiate all the different HashMap types floating around to make it more clear what is being consumed and returned.

How does it look now? A step in the right direction or should I revert?

Reply:
Thanks. I've added a couple of nits as inline comments.

I'm not sure about `PartialResults`. It might be useful to pass around global data as `&mut PartialResults` (see also inline comments about temporary `HashMap`), but that might be unnecessary if we can simplify and inline some of the functions.